### PR TITLE
Refactor DFA.minify()

### PR DESCRIPTION
**Pull request by eliotwrobson**
@caleb531 Refactors the minify and cross product functions to avoid creating intermediate DFAs. Cross product function now returns only reachable part of the product DFA by default.

### CODE REVIEW

**Review by caleb531 — CHANGES_REQUESTED** [review]
@eliotwrobson Looks good so far, although there are a few `filter` calls that I would prefer to have written as set comprehensions.

File: automata/fa/dfa.py [hunk 1]
```diff
@@ -377,65 +383,115 @@ def union(self, other, *, retain_names=False, minify=True):
...
+        new_states, new_transitions, new_initial_state = self._cross_product(other)
 
         if minify:
-            return new_dfa.minify(retain_names=retain_names)
+            return self._minify(
+                reachable_states=new_states,
+                input_symbols=self.input_symbols,
+                transitions=new_transitions,
+                initial_state=new_initial_state,
+                reachable_final_states=set(filter(union_fn, new_states)),
```
[the remark this answers is not in the document]
Comment (eliotwrobson) [reply automata/fa/dfa.py]: @caleb531 done!

**Comment by eliotwrobson** [comment]
@caleb531 needs a rebase after merging my other PR, but should be good to go otherwise.

**Review by Tagl — APPROVED** [review]
Looks good

File: automata/fa/dfa.py [hunk 2]
```diff
@@ -626,7 +704,7 @@ def compress(word, next_word):
         )
 
     @classmethod
-    def from_nfa(cls, target_nfa, retain_names=False):
+    def from_nfa(cls, target_nfa, *, retain_names=False, minify=True):
```
Comment (caleb531) [inline automata/fa/dfa.py:707]: @eliotwrobson Just want to say good idea to make these keyword-only arguments 🙂

**Comment by caleb531** [comment]
@eliotwrobson This looks great! Will approve and merge now. Thank you for this!

**Review by caleb531 — APPROVED** [review]

**Comment by caleb531** [comment]
@eliotwrobson I noticed there's a partial branch on one of the `_minify` lines when I run the coverage report:

Screen Shot 2022-10-25 at 4 28 53 PM

To solve this (i.e. bring the coverage back to 100%), is this just as simple as me minifying a DFA with at least one unreachable state?

**Comment by eliotwrobson** [comment]
@caleb531 I think that's correct. Now that the other functions that were calling minify only generate reachable states by default, it's not getting hit. You can probably just add an extra dummy state to one of the tests for minify.

**Pull request merged**